New rule/deployment outer scope evaluation

ms-sambell opened the pull request:
## PR Summary

Adding a new rule to look at Deployment resources that are set to use the `outer` scope expression in combination with a secureString or secureObject parameter. This can cause the parameter to be stored in clear text when the template is rendered.

Fixes: https://github.com/Azure/PSRule.Rules.Azure/issues/1475

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [x] Unit tests created/ updated
  - [x] Rule documentation created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [ ] Unit tests created/ updated
  - [ ] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

ms-sambell commented:
@BernieWhite Just wanted to check in with you on this Rule, I know it's not passing the test right now, but I was hoping to rely on the fix that'll come once this bug has been resolved: https://github.com/Azure/PSRule.Rules.Azure/issues/1687. 

My understanding is that once this is done, I can update the assertion to just check the deployments resources for the secure replacement text.

BernieWhite commented on docs/en/rules/Azure.Deployment.OuterSecret.md:
```diff
@@ -0,0 +1,90 @@
+
+## RECOMMENDATION
+
+Consider scoping the deployment to `inner` instead.
```
```suggestion
Consider using `inner` deployments to prevent secure values from being exposed.
```

BernieWhite commented on docs/en/rules/Azure.Deployment.OuterSecret.md:
```diff
@@ -0,0 +1,90 @@
+
+## DESCRIPTION
+
+Template child deployments can be scoped as either `outer` or `inner`. The `outer` scope uses parameters from the parent
+template which can be render into clear text. This exposes the secure information.
```
```suggestion
Template child deployments can be scoped as either `outer` or `inner`.
When using `outer` scope evaluated deployments, parameters from the parent template are used directly within nested templates instead of enforcing `secureString` and `secureObject` types.

When passing secure values to nested deployments always use `inner` scope deployments to ensure secure values are not logging.

Bicep modules always use `inner` scope evaluted deployments.
```

BernieWhite commented on src/PSRule.Rules.Azure/rules/Azure.Deployment.Rule.ps1:
```diff
@@ -55,4 +55,25 @@ function global:RecurseDeploymentSensitive {
+Rule 'Azure.Deployment.OuterSecret' -Ref 'AZR-000314' -Type 'Microsoft.Resources/deployments' -Tag @{ release = 'GA'; ruleSet = '2022_12' } { 
+    foreach($deployments in $TargetObject.properties.template.resources){
+        if($deployments.properties.expressionEvaluationOptions.scope -eq 'outer'){
+            Write-Host "Scope matched outer"
```
Use `Write-Verbose` or `Write-Debug` instead of `Write-Host`.

BernieWhite reviewed (changes requested):
@ms-sambell Nice. Have a look at the comments below.

ms-sambell commented:
Hey @BernieWhite, 

I'm changing the approach for this rule to:

1. Look at the ARM template to check if there's an outer deployment. 
2. Look up the rendered template in the Export-Template file and check if the parameter is a `reference`.

From what I observed in the Export-Template file the `secretReference` only shows outside of the `armTemplate` so my initial attempt was never going to work. 

Outer Deployment example:

```
{
        "ResourceName": "C:\\Users\\sambell\\code\\github\\PSRule.Rules.Azure\\tests\\PSRule.Rules.Azure.Tests\\Resources.Deployments.Parameters.json",
        "name": "nestedDeployment-D",
        "properties": {
            "template": {
                "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
                "contentVersion": "1.0.0.0",
                "parameters": {},
                "variables": {},
                "resources": [
                    {
                        "apiVersion": "2019-12-01",
                        "type": "Microsoft.Compute/virtualMachines",
                        "name": "vm-example",
                        "location": "australiaeast",
                        "properties": {
                            "osProfile": {
                                "computerName": "vm-example",
                                "adminUsername": "[parameters('adminUsername')]"
                            }
                        }
                    }
                ]
            },
            "templateLink": {
                "id": "/subscriptions/ffffffff-ffff-ffff-ffff-ffffffffffff/resourceGroups/ps-rule-test-rg",
                "uri": "https://deployment-uri"
            },
            "parameters": {
                "adminUsername": {
                    "reference": {
                        "keyVault": {
                            "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/rg-test/providers/Microsoft.KeyVault/vaults/keyvault-A"
                        },
                        "secretName": "string"
                    }
                },
                "adminObject": {
                    "reference": {
                        "keyVault": {
                            "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/rg-test/providers/Microsoft.KeyVault/vaults/keyvault-A"
                        },
                        "secretName": "object"
                    }
                }
            },
            "mode": "Incremental",
            "provisioningState": "Accepted",
            "templateHash": "52616168",
            "outputs": {}
        },
        "location": "eastus",
        "type": "Microsoft.Resources/deployments",
        "metadata": null,
        "_PSRule": {
            "path": "resources[3]",
            "source": [
                {
                    "file": "C:\\Users\\sambell\\code\\github\\PSRule.Rules.Azure\\tests\\PSRule.Rules.Azure.Tests\\Resources.Deployments.Template.json",
                    "type": "Template"
                },
                {
                    "file": "C:\\Users\\sambell\\code\\github\\PSRule.Rules.Azure\\tests\\PSRule.Rules.Azure.Tests\\Resources.Deployments.Parameters.json",
                    "type": "Parameter"
                }
            ]
        }
    }
```

See how the parameter type here contains secretName.

Inner Deployment example:

```
{
        "ResourceName": "C:\\Users\\sambell\\code\\github\\PSRule.Rules.Azure\\tests\\PSRule.Rules.Azure.Tests\\Resources.Deployments.Parameters.json",
        "name": "nestedDeployment-B",
        "properties": {
            "template": {
                "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
                "contentVersion": "1.0.0.0",
                "parameters": {
                    "adminObject": {
                        "type": "secureObject"
                    }
                },
                "variables": {},
                "resources": [
                    {
                        "apiVersion": "2019-12-01",
                        "type": "Microsoft.Compute/virtualMachines",
                        "name": "vm-example",
                        "location": "australiaeast",
                        "properties": {
                            "osProfile": "[parameters('adminObject')]"
                        }
                    }
                ]
            },
            "templateLink": {
                "id": "/subscriptions/ffffffff-ffff-ffff-ffff-ffffffffffff/resourceGroups/ps-rule-test-rg",
                "uri": "https://deployment-uri"
            },
            "parameters": {
                "adminObject": {
                    "value": "{{SecretReference:object}}"
                }
            },
            "mode": "Incremental",
            "provisioningState": "Accepted",
            "templateHash": "36390618",
            "outputs": {}
        },
        "location": "eastus",
        "type": "Microsoft.Resources/deployments",
        "metadata": null,
        "_PSRule": {
            "path": "resources[1]",
            "source": [
                {
                    "file": "C:\\Users\\sambell\\code\\github\\PSRule.Rules.Azure\\tests\\PSRule.Rules.Azure.Tests\\Resources.Deployments.Template.json",
                    "type": "Template"
                },
                {
                    "file": "C:\\Users\\sambell\\code\\github\\PSRule.Rules.Azure\\tests\\PSRule.Rules.Azure.Tests\\Resources.Deployments.Parameters.json",
                    "type": "Parameter"
                }
            ]
        }
    }
```

The parameter is set as the secretReference value, which means PSRule replaced the value. 


So my understanding is, that InnerLooped templates will have the `secretReference` replacement in their templates, where OuterLooped template will have parameters with the secret type in them.

ms-sambell commented:
Hey @BernieWhite, 

This rule isn't complete. The bit I was hoping for feedback on:

- The $TargetObject value doesn't contain the rendered parameters, so I can't use anything to match on. I can see there are replacements going on in the Export-Template output. The `cleanValue` is just a function string during the assessment. 

Is there something logical I can do here? Can I reference a different subsection in the Export-Template file during the rule evaluation? 

I was thinking something like: checking if there's a deployment with scope: outer and lookup the exported template, so I can confirm the value.

BernieWhite reviewed (commented):
Checkout the latest updates from main. I think you could use the `HasSecureValue` similar to the following but as a negative check.

https://github.com/Azure/PSRule.Rules.Azure/blob/c8a9116c9a5e898c2ef85b8ea8c294a70dfdfd48/src/PSRule.Rules.Azure/rules/Azure.Deployment.Rule.ps1#L63-L90

ms-sambell commented:
@BernieWhite - This should be functional now. 

I added an if condition to the rule evaluation `(-Type 'Microsoft.Resources/deployments' -If { IsParentDeployment })`. 

The rule was picking up both the template and the nested templates as different objects. The nested deployment objects didn't include the deployment scope. So there was no way to tell if they were scoped to outer. So I excluded them instead. 

I've also added the nested looping as discussed.

BernieWhite reviewed (approved):
@ms-sambell Great work! Updates look good.

There is a few transient ADO pipeline issues at the moment. Might need to rerun the pipeline tomorrow, otherwise all good to merge.
